## Lý do chọn thuật toán Gaussian Naive Bayes

**Gaussian Naive Bayes (GNB)** được lựa chọn vì các lý do sau:

- **Phù hợp với dữ liệu dạng liên tục:**  
  Bộ dữ liệu *Pima Indians Diabetes Database* có các thuộc tính như `Glucose`, `BloodPressure`, `BMI`,… là các giá trị **số thực (continuous)**.  
  Thuật toán GNB giả định rằng dữ liệu trong mỗi lớp tuân theo **phân phối chuẩn (Gaussian)**, nên rất phù hợp cho loại dữ liệu này.

- **Đơn giản, dễ huấn luyện và giải thích:**  
  GaussianNB dựa trên **định lý Bayes** và giả định **độc lập giữa các đặc trưng**, giúp việc tính toán xác suất nhanh, hiệu quả và dễ hiểu — đặc biệt quan trọng trong bài toán y học cần mô hình minh bạch.

- **Hiệu quả với dữ liệu nhỏ và vừa:**  
  Bộ dữ liệu có **1268 mẫu** và **8 đặc trưng**, do đó GNB huấn luyện nhanh, không cần tinh chỉnh nhiều siêu tham số.

- **Phù hợp bài toán phân loại nhị phân:**  
  Biến mục tiêu `Outcome` gồm:
  - `0`: Không bị tiểu đường  
  - `1`: Bị tiểu đường  
  ⇒ Đây là **bài toán phân loại nhị phân (binary classification)** — rất thích hợp cho GNB.

---

## 📊 Chia tập train/test

Sau khi xử lý giá trị thiếu, dữ liệu được chia thành hai phiên bản:

| Dataset | Cách xử lý giá trị thiếu |
|----------|--------------------------|
| **Dataset 1** | Thay NaN bằng **Mean** |
| **Dataset 2** | Thay NaN bằng **Median** |

Mỗi tập dữ liệu được chia thành:
- **80% train set**: dùng để huấn luyện mô hình  
- **20% test set**: dùng để đánh giá mô hình  
- Dùng `random_state=42` để tái lập kết quả.

```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)


In [5]:
# ===========================
# 03b_GaussianNB_Training.ipynb
# ===========================

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, precision_score, recall_score, f1_score

# ===========================
# ĐỌC DỮ LIỆU
# ===========================
df1 = pd.read_csv("../../datasets/processed/diabetes_dataset1_mean.csv")
df2 = pd.read_csv("../../datasets/processed/diabetes_dataset2_median.csv")


def evaluate_model(y_true, y_pred, model_name, dataset_name):
    print(f"\n{'='*60}")
    print(f"{model_name} - {dataset_name}")
    print(f"{'='*60}")

    acc = accuracy_score(y_true, y_pred)
    prec = precision_score(y_true, y_pred)
    rec = recall_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred)

    print(f"- Accuracy:  {acc:.4f}")
    print(f"- Precision: {prec:.4f}")
    print(f"- Recall:    {rec:.4f}")
    print(f"- F1-Score:  {f1:.4f}")
    print("\nConfusion Matrix:")
    print(confusion_matrix(y_true, y_pred))
    print("\nClassification Report:")
    print(classification_report(y_true, y_pred, target_names=['Không bị', 'Bị']))

    # return acc, prec, rec, f1

# ===========================
# TRAIN & TEST GaussianNB
# ===========================
print("\n--- Gaussian NB với Dataset 1 (Mean) ---")
X1 = df1.drop('Outcome', axis=1)
y1 = df1['Outcome']
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.2, random_state=42)
nb1 = GaussianNB()
nb1.fit(X1_train, y1_train)
y1_pred = nb1.predict(X1_test)
evaluate_model(y1_test, y1_pred, "Gaussian NB", "Dataset 1 (Mean)")

print("\n--- Gaussian NB với Dataset 2 (Median) ---")
X2 = df2.drop('Outcome', axis=1)
y2 = df2['Outcome']
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.2, random_state=42)
nb2 = GaussianNB()
nb2.fit(X2_train, y2_train)
y2_pred = nb2.predict(X2_test)
evaluate_model(y2_test, y2_pred, "Gaussian NB", "Dataset 2 (Median)")



--- Gaussian NB với Dataset 1 (Mean) ---

Gaussian NB - Dataset 1 (Mean)
- Accuracy:  0.7402
- Precision: 0.6034
- Recall:    0.4487
- F1-Score:  0.5147

Confusion Matrix:
[[153  23]
 [ 43  35]]

Classification Report:
              precision    recall  f1-score   support

    Không bị       0.78      0.87      0.82       176
          Bị       0.60      0.45      0.51        78

    accuracy                           0.74       254
   macro avg       0.69      0.66      0.67       254
weighted avg       0.73      0.74      0.73       254


--- Gaussian NB với Dataset 2 (Median) ---

Gaussian NB - Dataset 2 (Median)
- Accuracy:  0.7559
- Precision: 0.6429
- Recall:    0.4615
- F1-Score:  0.5373

Confusion Matrix:
[[156  20]
 [ 42  36]]

Classification Report:
              precision    recall  f1-score   support

    Không bị       0.79      0.89      0.83       176
          Bị       0.64      0.46      0.54        78

    accuracy                           0.76       254
   macro av

In [6]:
import joblib
joblib.dump(nb1, "../../models/diabetes_nb_model.pkl")
print("✅ Mô hình đã được lưu thành công!")

✅ Mô hình đã được lưu thành công!


## 🔹 ĐÁNH GIÁ VÀ NHẬN XÉT KẾT QUẢ MÔ HÌNH GAUSSIAN NAIVE BAYES (GNB)

### 🧮 1. Kết quả huấn luyện và kiểm thử

#### **Dataset 1 (Thay giá trị thiếu bằng Mean)**
- **Accuracy:** 0.7402  
- **Precision:** 0.6034  
- **Recall:** 0.4487  
- **F1-Score:** 0.5147  

**Ma trận nhầm lẫn:**
|            | Dự đoán Không bị | Dự đoán Bị |
|-------------|------------------|-------------|
| **Thực tế Không bị** | 153 | 23 |
| **Thực tế Bị**        | 43 | 35 |

---

#### **Dataset 2 (Thay giá trị thiếu bằng Median)**
- **Accuracy:** 0.7559  
- **Precision:** 0.6429  
- **Recall:** 0.4615  
- **F1-Score:** 0.5373  

**Ma trận nhầm lẫn:**
|            | Dự đoán Không bị | Dự đoán Bị |
|-------------|------------------|-------------|
| **Thực tế Không bị** | 156 | 20 |
| **Thực tế Bị**        | 42 | 36 |

---

### 📊 2. Phân tích kết quả mô hình

Mô hình **Gaussian Naive Bayes (GNB)** là thuật toán xác suất dựa trên **Định lý Bayes**, với giả định rằng các thuộc tính **độc lập có điều kiện theo từng lớp**. Nhờ đặc điểm này, GNB có ưu điểm về **tốc độ huấn luyện nhanh** và **cấu trúc đơn giản**, phù hợp với các bộ dữ liệu y học có kích thước vừa như **Pima Indians Diabetes Dataset**.

Kết quả thực nghiệm thu được như sau:

- **Dataset 1 (Mean)**  
  - Accuracy: **74.02%**  
  - Precision: **60.34%**  
  - Recall: **44.87%**  
  - F1-score: **51.47%**

- **Dataset 2 (Median)**  
  - Accuracy: **75.59%**  
  - Precision: **64.29%**  
  - Recall: **46.15%**  
  - F1-score: **53.73%**

Kết quả cho thấy mô hình đạt **độ chính xác khoảng 74–76%**, tuy nhiên khả năng phát hiện đúng các trường hợp **bị tiểu đường** vẫn còn hạn chế.

---

### 🔍 3. Nhận xét chi tiết

**1. Khả năng dự đoán**  
Mô hình dự đoán tốt các trường hợp **không bị tiểu đường**, thể hiện qua Recall cao (87–89%).  
Ngược lại, Recall của lớp **bị tiểu đường** chỉ đạt khoảng **45–46%**, cho thấy mô hình vẫn bỏ sót một số ca bệnh thực tế.

**2. So sánh dữ liệu Mean và Median**  
Việc xử lý dữ liệu bằng **Median** cho kết quả tốt hơn nhẹ so với Mean:
- Accuracy tăng từ **74.02%** lên **75.59%**
- F1-score tăng từ **0.51** lên **0.54**

Điều này cho thấy Median giúp **giảm ảnh hưởng của các giá trị ngoại lai** trong dữ liệu.

**3. Phân tích ma trận nhầm lẫn**  
- Dataset Mean: mô hình dự đoán đúng **153** trường hợp không bị và **35** trường hợp bị tiểu đường.  
- Dataset Median: số trường hợp dự đoán đúng lớp bệnh tăng lên **36**, đồng thời số dự đoán sai giảm nhẹ.

Mô hình có xu hướng **thiên về lớp đa số**, đây là hạn chế phổ biến của Gaussian Naive Bayes.

**4. Đặc điểm mô hình**  
Giả định các thuộc tính độc lập của GNB không hoàn toàn phù hợp với dữ liệu y học, tuy nhiên mô hình vẫn đạt kết quả chấp nhận được, cho thấy **khả năng tổng quát hóa tốt**.

---

### ⚙️ 4. Tổng kết

- Gaussian Naive Bayes đạt độ chính xác **74–76%** trên bộ dữ liệu Pima.  
- F1-score của lớp bị tiểu đường đạt khoảng **0.51–0.54**.  
- Mô hình **đơn giản, dễ triển khai, tốc độ huấn luyện nhanh**, phù hợp làm **mô hình baseline**.

---

### 🧠 Kết luận

Gaussian Naive Bayes là lựa chọn phù hợp khi cần một mô hình **nhanh và nhẹ** cho bài toán dự đoán tiểu đường.  
Mặc dù khả năng phát hiện lớp bệnh còn hạn chế, mô hình vẫn đóng vai trò quan trọng làm nền tảng để so sánh và phát triển các phương pháp học máy nâng cao hơn.
